### ISSUE TITLE
Quick version string hack that was never fixed

### ISSUE THREAD

**Issue by dongcarl** [bitcoin/bitcoin#16588]
In #11097 we added a quick version string hack that seems to do the following:

-----

When Gitian building, ***after*** generating the source tarball (a la `make dist`), we invoke `make -C src obj/build.h`, as `make dist` does not: https://github.com/bitcoin/bitcoin/blob/b499d8576f30d13dfe9000852afe47ee1c28cfbe/src/Makefile.am#L245-L248
We can see that this calls `share/genbuild.sh`, which generates a `src/obj/build.h` containing version information obtained from `git`: https://github.com/bitcoin/bitcoin/blob/b499d8576f30d13dfe9000852afe47ee1c28cfbe/share/genbuild.sh#L1-L53

Then, for each architecture we build, we neuter `share/genbuild.sh` (as normal `make` will try to regenerate `obj/build.h`, but now lacks the build information as we're not in a git repository anymore) and copy over the generated `src/obj/build.h` mentioned above. All of this is being done before `configure`ing, `make`ing, and `make install`ing for the architecture: https://github.com/bitcoin/bitcoin/blob/b499d8576f30d13dfe9000852afe47ee1c28cfbe/contrib/gitian-descriptors/gitian-linux.yml#L168-L177

-----

Perhaps what we can do is simply add `src/obj/build.h` to `EXTRA_DIST`, and generate it in a `dist-hook`? Do non-existent files in `EXTRA_DIST` get their targets called? Source: https://www.gnu.org/software/automake/manual/html_node/The-dist-Hook.html

@theuni @luke-jr would love some more context here.

**Comment by MarcoFalke** [comment]
Can this be moved to the `make dist` stage and also made to fail when no git is available? I think that makes most sense.

Is it supported to call `make dist` recursively? I guess not, so we can probably safely assume that `git` is always available at build time.

**Comment by hebasto** [comment]
> Perhaps what we can do is simply add `src/obj/build.h` to `EXTRA_DIST`, and generate it in a `dist-hook`?

Will gitian builds have `CLIENT_BUILD` set correctly?

**Comment by luke-jr** [comment]
make will try to update it if it thinks the timestamps are meeting some criteria.

I think we want to just disable the rule to generate build.h if there's no git info...

**Comment by dongcarl** [comment]
Some discussion in another thread: https://github.com/bitcoin/bitcoin/pull/16611#issuecomment-521396904

**Comment by MarcoFalke** [comment]
If I build Bitcoin Core with `make distdir VERSION=master && cd bitcoin-master` I get:

```
2019-08-23T17:36:07Z Bitcoin Core version v0.18.99.0-unk (debug build)
```

I presume (hope) this is not the case when building from the gitian produced distribution tar. Regardless, this sounds like a bug that should be fixed.

**Comment by fanquake** [comment]
@dongcarl I know GitHub has autoclosed this. Can you confirm that there's nothing left to do here after #18556.

**Comment by dongcarl** [comment]
I think this is fixed right now, and as long as we don't generate anything during the release build process that would make `git diff-index --quiet HEAD --` return a non-zero status code, we're good.

My read of the current behaviour:

1. When generating the release tarball, we no longer generate `src/obj/build.h` as of 1362be044724bb49d785ca2e296a3b43343c1690
2. When building for each `$HOST` under `distsrc-$HOST`, `share/genbuild.sh` is called, and the `git` commands inside it will traverse up until it hits the top git repo (effectively `distsrc-$HOST/../`), and the commands will work.

A corner case that might be confusing: when we perform a build on a dirty repo, the release tarball will be of `HEAD` (let's say that's commit hash `fffffffff`). HOWEVER, when `share/genbuild.sh` is being run under `distsrc-$HOST`, it will output something like

```c
#define BUILD_SUFFIX fffffffff-dirty
```

When really the `distsrc-$HOST` is not dirty at all.

I don't know if that belongs in/deserves a separate issue or not
